fdbserver/ptxn/MessageSerializer.cpp

```diff
@@ -0,0 +1,347 @@
+	ASSERT(subsequenceMutationItem.subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequenceMutationItem.subsequence;
+	serializer.writeItem(subsequenceMutationItem);
+}
+
+void SubsequencedMessageSerializer::write(const Subsequence& subsequence, const MutationRef& mutation) {
+	ASSERT(subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequence;
+	serializer.writeItem(SubsequenceMutationItem{ subsequence, mutation });
+}
+
+void SubsequencedMessageSerializer::write(const SubsequenceSpanContextItem& subsequenceSpanContextItem) {
+	ASSERT(subsequenceSpanContextItem.subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequenceSpanContextItem.subsequence;
+	serializer.writeItem(subsequenceSpanContextItem);
+}
+
+void SubsequencedMessageSerializer::write(const Subsequence& subsequence, const SpanContextMessage& spanContext) {
+	ASSERT(subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequence;
+	serializer.writeItem(SubsequenceSpanContextItem{ subsequence, spanContext });
+}
+
+void SubsequencedMessageSerializer::write(const SubsequenceLogProtocolMessageItem& subsequenceLogProtocolMessageItem) {
+	ASSERT(subsequenceLogProtocolMessageItem.subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequenceLogProtocolMessageItem.subsequence;
+	serializer.writeItem(subsequenceLogProtocolMessageItem);
+}
+
+void SubsequencedMessageSerializer::write(const Subsequence& subsequence,
+                                          const LogProtocolMessage& logProtocolMessage) {
+	ASSERT(subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequence;
+	serializer.writeItem(SubsequenceLogProtocolMessageItem{ subsequence, logProtocolMessage });
+}
+
+void SubsequencedMessageSerializer::write(
+    const Subsequence& subsequence,
+    const std::variant<MutationRef, SpanContextMessage, LogProtocolMessage>& message) {
+
+	switch (static_cast<MutationRef::Type>(message.index())) {
+	case MutationRef::Reserved_For_SpanContextMessage:
+		write(subsequence, std::get<SpanContextMessage>(message));
+		break;
+	case MutationRef::Reserved_For_LogProtocolMessage:
+		write(subsequence, std::get<LogProtocolMessage>(message));
+		break;
+	default:
+		write(subsequence, std::get<MutationRef>(message));
+		break;
+	}
+}
+
+const Version& SubsequencedMessageSerializer::getCurrentVersion() const {
+	return header.lastVersion;
+}
+
+const Subsequence& SubsequencedMessageSerializer::getCurrentSubsequence() const {
+	return header.lastSubsequence;
+}
+
+void SubsequencedMessageSerializer::completeVersionWriting() {
+	// If a version exists, yet has 0 items (an empty commit), a NOOP is injected to avoid version skipping. Since the
+	// NOOP basically does nothing, it has the maximum subsequence number, which *hopefully* will never be reached.
+	if (serializer.getNumItemsCurrentSection() == 0) {
+		write(MAX_SUBSEQUENCE, MutationRef{ MutationRef::NoOp, ""_sr, ""_sr });
```

Comment:
Why does it need to inject a mutation? Could this version just be empty? This will complicates the deserializer to perform special handling?

Reply:
After some thoughts, I feel I would like to inject a mutation here. The reason is that, it is hard to trace a Version with no mutations followed up. This *might* cause some difficulties when we are trying to ensure the continuous of the versions during persistence. With a Noop, the version is preserved.

Reply:
I don't understand the reasoning here. Conceptually, it's basically a version with empty data. Physically, it could be 0 items for a version. Persistence is not related to serialization: tlog or storage server track versions by themselves and have no business with how serialization is implemented.

Maybe you are having difficulty with the deserialization interface (i.e., iterator), which doesn't allow empty data for a version?

Reply:
The iterator will bypass the empty data. I am having the hesitation of keeping the versions continuous during the storage process. The reason is that I think if we persist the data in serialized format, TLog might or might not need such information. Let me first remove it and, if necessary, add it back.